script.js

```diff
@@ -0,0 +1,942 @@
+      } else if (end !== text.value.length) {
+        // handle out of right boundaries
+        text.selectionStart = start + 1;
+        text.selectionEnd = text.selectionStart;
+      }
+      break;
+    }
+    default: {
+      if (!isMouse() && (event.ctrlKey || event.shiftKey || event.altKey)) {
+        break;
+      }
+
+      if (isMouse()) {
+        event.preventDefault();
+        const start = text.selectionStart;
+        const end = text.selectionEnd;
+
+        const strBefore = text.value.substring(0, start);
+        const strAfter = text.value.substring(end);
+        const symbol = el.innerText;
+
+        // set textarea value to: text before caret + tab + text after caret
+        text.value = `${strBefore}${symbol}${strAfter}`;
+
+        // put caret at right position
+        text.selectionStart = start + symbol.length;
+        text.selectionEnd = text.selectionStart;
+      }
+      break;
+    }
+  }
+}
+
+const isSpecialKey = (text, e) => (
+  (text === variables.specialKeys.ESC && e.key === 'Escape')
+|| (text === variables.specialKeys.SPACE && e.code === 'Space')
+|| (text === variables.specialKeys.ARROW_UP && e.code === 'ArrowUp')
+|| (text === variables.specialKeys.ARROW_DOWN && e.code === 'ArrowDown')
+|| (text === variables.specialKeys.ARROW_LEFT && e.code === 'ArrowLeft')
+|| (text === variables.specialKeys.ARROW_RIGHT && e.code === 'ArrowRight')
+|| (text === variables.specialKeys.PAGE_UP && e.code === 'PageUp')
+|| (text === variables.specialKeys.PAGE_DOWN && e.code === 'PageDown')
+|| (text === variables.specialKeys.ALTGR && e.code === 'AltRight')
+|| (text === variables.specialKeys.CTRL && e.code === 'ControlLeft')
+|| (text === variables.specialKeys.SHIFT && e.code === 'ShiftLeft')
+|| (text === variables.specialKeys.SCROLL_LOCK && e.key === 'ScrollLock')
+|| (text === variables.specialKeys.PRINT_SCREEN && e.key === 'PrintScreen')
+|| (text === '<' && e.code === 'IntlBackslash')
+);
+
+const isSecondKey = (text, e) => (
+  (text === variables.specialKeys.CTRL && e.code === 'ControlRight')
+|| (text === variables.specialKeys.SHIFT && e.code === 'ShiftRight')
+|| (text === variables.specialKeys.ALT && e.code === 'AltRight')
+|| (text === variables.specialKeys.ENTER && e.code === 'NumpadEnter')
+|| (text === '\\' && e.code === 'IntlBackslash')
+|| (text === '/' && e.code === 'Slash')
+|| (text === '-' && e.code === 'NumpadSubtract')
+|| (text === '0' && e.code === 'Numpad0')
+|| (text === '1' && e.code === 'Numpad1')
+|| (text === '2' && e.code === 'Numpad2')
+|| (text === '3' && e.code === 'Numpad3')
+|| (text === '4' && e.code === 'Numpad4')
+|| (text === '5' && e.code === 'Numpad5')
+|| (text === '6' && e.code === 'Numpad6')
+|| (text === '7' && e.code === 'Numpad7')
+|| (text === '8' && e.code === 'Numpad8')
+|| (text === '9' && e.code === 'Numpad9')
+|| (text === '.' && e.code === 'NumpadDecimal')
+);
+
+const processKeySelection = (e) => {
+  let isKeydown;
+  switch (e.type) {
```

Comment:
Empty line between declaration and switch is needed. For code style you can use eslint or prettier. It will be cool if you configure it and create some basic rules.

Reply:
I couldn't find config between declaration and switch, need to find it out later